syntax = "proto2";


message Seq2SeqModel {
  optional PredictionModel prediction_model = 1;
  optional Dataset dataset = 2;
  optional Decoding decoding = 3;
  optional Optimization optimization = 5;
  optional int64 src_vocab_size = 6;
  optional int64 tgt_vocab_size = 7;
  optional int64 random_seed = 8;
  optional int64 steps_per_stats = 9 [default = 100];
  optional int64 steps_ckpt = 10 [default = 1000];
}

message PredictionModel {
  optional string unit_type = 1 [default = "lstm"];
  optional int64 num_units = 2 [default = 512];
  optional float forget_bias = 3 [default = 1.0];
  optional float keep_prob = 4 [default = 0.8];
  optional string encoder_type = 5 [default = "bi"];
  optional bool time_major = 6 [default = true];
  optional bool share_vocab = 7 [default = false];
  optional Initializer initializer = 9;
  optional int64 num_encoder_layers = 12 [default = 2];
  optional int64 num_encoder_res_layers = 13 [default = 0];
  optional int64 num_decoder_layers = 14 [default = 2];
  optional int64 num_decoder_res_layers = 15 [default = 0];  
  optional string attention_type = 16 [default = 'scaled_luong'];
  optional bool output_attention = 17 [default = true];
}

message Dataset {
  optional int64 batch_size = 1 [default = 128];
  optional int64 shuffle_buffer_size = 2 [default = 10000];
  optional int64 num_buckets = 3 [default = 5];
  optional int64 src_max_len = 4 [default = 50];
  optional int64 tgt_max_len = 5 [default = 50];
  optional string sos = 6 [default = "<s>"];
  optional string eos = 7 [default = "</s>"];
}

message Decoding {
  optional int64 beam_width = 1 [default = 10];
  optional float length_penalty_weight = 2 [default = 0.0];
  optional float sampling_temperature = 3 [default = 1.0];
  optional int64 maximum_iterations = 4 [default = 50];
}

message Initializer {
  message RandomUniformInitializer {
    optional float min_val = 1 [default = -0.1];
    optional float max_val = 2 [default = 0.1];
  }

  message TruncatedNormalInitializer {
    optional float mean = 1 [default = 0.0];
    optional float stddev = 2 [default = 0.1];
  }

  oneof initializer_oneof {
    RandomUniformInitializer uniform = 1;
    TruncatedNormalInitializer truncated_normal = 2;
  }
}

message Optimization {
  optional float base_learning_rate = 1 [default = 1.0];
  optional string warmup_scheme = 2 [default = "t2t"];
  optional int64 warmup_steps = 3 [default = 0];
  optional string decay_scheme = 4 [default = 'luong234'];
  optional string optimizer = 5 [default = 'sgd'];
  optional int64 num_train_steps = 6;
  optional float max_grad_norm = 7 [default = 5.0]; 
}

